lib/vsdata.cpp

```diff
@@ -47,6 +47,104 @@ void VisualizationSceneScalarData::FixValueRange()
    }
 }
 
+int VisualizationSceneScalarData::GetFunctionAutoRefineFactor(GridFunction &gf)
+{
+   Mesh *mesh = gf.FESpace()->GetMesh();
+   const int dim = mesh->Dimension();
+   int ref = 1;
+   if (dim == 3)
+   {
+      for (int i = 0; i < mesh->GetNBE(); i++)
+      {
+         const int f = mesh->GetBdrElementFaceIndex(i);
+         const FiniteElementCollection *fec = gf.FESpace()->FEColl();
+         const FiniteElement *fe = NULL;
+         if (fec && fec->GetContType() != FiniteElementCollection::DISCONTINUOUS)
+         {
+            fe = gf.FESpace()->GetFaceElement(f);//might abort instead of returning NULL!
+         }
+         int order;
+         if (fe)
+         {
+            order = fe->GetOrder();
+            if (fe->GetMapType() == FiniteElement::MapType::INTEGRAL)
+            {
+               order += mesh->GetBdrElementTransformation(i)->OrderW();
+            }
+         }
+         else
+         {
+            int ivol, info;
+            mesh->GetBdrElementAdjacentElement(i, ivol, info);
+            const FiniteElement *fevol = gf.FESpace()->GetFE(ivol);
+            order = fevol->GetOrder() - 1;
+            if (fevol->GetMapType() == FiniteElement::MapType::INTEGRAL)
+            {
+               order += mesh->GetElementTransformation(ivol)->OrderW();
+            }
+         }
+         ref = std::max(ref, 2 * order);
+      }
+   }
+   else
+   {
+      for (int i = 0; i < mesh->GetNE(); i++)
+      {
+         const FiniteElement &fe = *gf.FESpace()->GetFE(i);
+         int order = fe.GetOrder();
+         if (fe.GetMapType() == FiniteElement::MapType::INTEGRAL)
+         {
+            order += mesh->GetElementTransformation(i)->OrderW();
```

Comment:
What is this? Definitely get rid of this.

Reply:
Integral elements should be more refined due to the mesh order. This is not precise, but at least something.

Reply:
I disagree, the order of the space should be sufficient.

Reply:
Even a zero-order integral finite element might have varying values due to the changing Jacobian 🤔

Reply:
Using `OrderW` is huge overkill -- on Q2 mesh in 3D it is 5. Anything beyond the max of the solution and mesh order is an overkill for me.

Reply:
But I know from Lagrangian MHD that this is an issue, just a bit squeezed element with integral elements (for magnetic field for example) even of the zero order gives ramping (real) values. I am open to different formulas, but it should be definitely higher than just plain order of the (value-based) element.

Reply:
You example is probably an exception than something typically observed. Even if you have curved mesh with piecewise constant (integral) field, then refinement factor will be more than 0 because of the mesh and that will be sufficient to indicate that there is variation within the cell; this initial subdivision can then be additionally increased, if desired, by the user.

Reply:
There is only a warning now, as discussed 😉 .